add cupertino localizations

doxuanviet1996 opened the pull request.

doxuanviet1996 committed:
add cupertino localizations

doxuanviet1996 committed:
add test

xster commented:
Is this ready for review? 

cc @HansMuller

xster commented on packages/flutter/lib/cupertino.dart:
```diff
@@ -13,6 +13,7 @@ export 'src/cupertino/app.dart';
 export 'src/cupertino/bottom_tab_bar.dart';
 export 'src/cupertino/button.dart';
 export 'src/cupertino/colors.dart';
+export 'src/cupertino/cupertino_localizations.dart';
```
Don't add 'cupertino_' again in the file names. The package already says cupertino. Just have the prefix in the class names, since those are the things that get re-exported.

HansMuller commented:
We're going to need to talk about this.  I expect that we'll want to factor out some of the material localizations implementation, notably all of the date stuff, into a shared library. Exactly how to factor things, what code generation tools to share, etc, warrants some thought.

doxuanviet1996 committed:
rename

doxuanviet1996 committed:
review fixes

doxuanviet1996 committed:
.

doxuanviet1996 commented on packages/flutter/lib/src/cupertino/localizations.dart:
````diff
@@ -0,0 +1,221 @@
+  /// Label that appears next to the hour picker in
+  /// [CupertinoCountdownTimerPicker] when selected hour value is ```hour```.
+  /// This function will deal with pluralization based on the ```hour``` parameter.
+  String timerPickerHourLabel(int hour);
````
I'm not sure if this is the best way to deal with pluralization. In English and many other common languages, there are at most 2 forms: singular and plural. However, that might not be true for all languages, so here I leave it as general as it can be, and let the overriding class deal with it.
What do you think @xster?

xster replied:
See the selectedRowCountTitle example
https://github.com/flutter/flutter/blob/master/packages/flutter_localizations/lib/src/l10n/README.md

doxuanviet1996 replied:
yup, the example has the same pattern.

xster replied:
Not sure if I answered the question. Having the a simpler DefaultCupertinoLocalization for english is fine. What does iOS do when it's 0?

xster commented on packages/flutter/lib/src/cupertino/localizations.dart:
```diff
@@ -0,0 +1,221 @@
+  /// Different languages can have different numeral systems, and this function
+  /// will return the given number written in the local language.
+  /// Example of some numeral systems can be found here:
+  /// https://en.wikipedia.org/wiki/List_of_numeral_systems
```
Flip side of the comments I was making yesterday:
Also make sure whatever API you create here is sufficient to satisfy the time picker use case. For instance, I suspect changing your phone to Chinese or Devanagari and using the time picker probably won't show
`〇/零 一/壹 二/貳 三/叄 四/䦉 五/伍 六/陸 七/柒 八/捌 九/玖 十/拾` or `0 १ २ ३ ४ ५ ६ ७ ८ ९`. 

Make sure you derive the requirement and then create APIs that match them.

xster commented on packages/flutter/lib/src/cupertino/localizations.dart:
````diff
@@ -0,0 +1,221 @@
+  String timerPickerHourLabel(int hour);
+
+  /// Label that appears next to the minute picker in
+  /// [CupertinoCountdownTimerPicker] when selected minute value is ```minute```.
````
single backtick

doxuanviet1996 committed:
doc + function name fixes

doxuanviet1996 committed:
add doc examples

doxuanviet1996 commented:
I have removed the general function `number` and replaced it with very specific one like `datePickerHour`, `datePickerMinute`, etc. I tried to find a common pattern and group them together in one function, but it does not seem to work.

doxuanviet1996 committed:
analyze fix

doxuanviet1996 committed:
indent fix

xster commented:
LGTM
